playwright/src/main/java/com/microsoft/playwright/impl/junit/BrowserContextExtension.java

```diff
@@ -90,7 +90,14 @@ public void testFailed(ExtensionContext extensionContext, Throwable cause) {
     if (shouldRecordTrace(options)) {
       saveTrace(extensionContext);
     }
+    cleanup(options);
+  }
+
+  private void cleanup(Options options) {
     closeBrowserContext();
+    if(options.closeBrowserAfterEachTest != null && options.closeBrowserAfterEachTest) {
+      BrowserExtension.closeBrowser();
```

Comment:
The behavior of closing the browser (rather than just the context) after each test is very specific to BrowserStack implementation and they could solve it better by providing an API for explicitly marking test start/end instead of binding it to the lifetime of the browser connection.

I'd rather the clients, that need to close browsers after each test, implemented their own fixture and just called `browser.close()` after each test from the fixture. We can add [`onDisconnect`](https://playwright.dev/java/docs/api/class-browser#browser-event-disconnected) handler to all browser instances that are created via `Playwright` fixture and clean-up thread local cache when the event is fired. It would also cover the case when the user calls `browser.close()` in the middle of a test too, so it's something that would make sense to do in any case.

Do you think such approach would make more sense or you'd still prefer to have the option?